resolver/resolver.go

```diff
@@ -467,28 +472,29 @@ func truncate(m *dns.Msg, udp bool) *dns.Msg {
 
 	furtherTruncation := m.Len() > max
 	m.Truncated = m.Truncated || furtherTruncation
-
+	if !setTruncateBit {
+		// Clear the Truncate bit regardless of whether this message used
+		// to have it set set or whether we are going to truncate it here.
+		m.Truncated = false
+	}
 	if !furtherTruncation {
 		return m
 	}
-
 	m.Extra = nil // Drop all extra records first
 	if m.Len() < max {
+		// Dropping the extra records shrunk the message down sufficiently.
 		return m
 	}
 	answers := m.Answer[:]
-	left, right := 0, len(m.Answer)
-	for {
-		if left == right {
-			break
-		}
-		mid := (left + right) / 2
-		m.Answer = answers[:mid]
-		if m.Len() < max {
-			left = mid + 1
-			continue
-		}
-		right = mid
+	search := func(n int) bool {
+		m.Answer = answers[:n]
+		return m.Len() > max
+	}
+	limit := sort.Search(len(answers), search)
+	if limit > 0 {
+		m.Answer = answers[:limit-1]
+	} else {
+		m.Answer = answers[:0]
```

Comment:
This doesn't blow away the array that backs `m.Answer`, which means `append(m.Answer , ...)` on the message returned from this function won't allocate a new backing array. 

This is a standard go idiom in high-performance environments. My impression is that mesos-dns is such an environment. Looking at this function in the greater context of the project this doesn't seem necessary as we never append answers to the message returned from truncate. Still, that is a global consideration that may change in the future so I'd rather write fast code here and not worry.